server/src/main/java/org/opensearch/search/query/QueryCollectorContext.java

```diff
@@ -67,38 +72,69 @@
  */
 @PublicApi(since = "1.0.0")
 public abstract class QueryCollectorContext {
-    private static final Collector EMPTY_COLLECTOR = new SimpleCollector() {
-        @Override
-        public void collect(int doc) {}
 
-        @Override
-        public ScoreMode scoreMode() {
-            return ScoreMode.COMPLETE_NO_SCORES;
-        }
-    };
+    private static Collector createEmptyCollector(ScoreMode scoreMode) {
```

Comment:
We have discovered this logic in MultiCollector while working on hybrid query. For the hybrid search scenario, we set the score mode to TOP_SCORES (code ref), but due to logic in MultiCollector, it falls back to COMPLETE. This leads to extensive collecting of scores and runtime exceptions (one example is described in https://github.com/opensearch-project/neural-search/issues/964).

The issue can be fixed in two ways - with a flexible QueryCollectorContext or in MultiCollector. The first approach introduced in this PR is simpler and is limited to scenarios when the score mode for all collectors in the multi collector is the same and is known beforehand. In this case, the client initializes an empty query context with any score mode that matches the one used in that client's logic. For example, for hybrid query, we will use TOP_SCORES (QueryCollectorContext.getContextForScoreMode(ScoreMode.TOP_SCORES)).

If any of these conditions are not met, then even with this change, the multi collector will still fall back to one of COMPLETE/COMPLETE_NO_SCORES, and the fix should be in the logic of MultiCollector.

Reply:
Thanks @martin-gaievski 

> If any of these conditions are not met, then even with this change, the multi collector will still fall back to one of COMPLETE/COMPLETE_NO_SCORES, and the fix should be in the logic of MultiCollector.

I think that is the place we have to fix, could you please open the issue / pr on Apache Lucene side? Thanks!

Reply:
Discussed with @reta offline, here is the summary:

1. We identified two scenarios: a "proper" multiple collectors case and an empty collector case.
2. For the immediate solution in OpenSearch core:
   - We'll add a constant collector specialization with TOP_SCORES mode, similar to the existing COMPLETE_NO_SCORES.
   - This approach is limited but sufficient for our current needs and can be easily removed later.
3. For the long-term solution:
   - I'll initiate a discussion on the Lucene side about a proper fix for MultiCollector.
   - We'll add a comment to the Lucene issue explaining our temporary approach in OpenSearch.
4. I'll adjust the PR to implement the constant collector specialization approach.

Reply:
> This leads to extensive collecting of scores and runtime exceptions

I suspect that the exception may still be a thing (but less of a thing) if we use `TOP_SCORES` instead of `COMPLETE`. The exceptions tend to suggest that there's a `TwoPhaseIterator` that's not being used properly -- that is, the approximation points to a doc that doesn't exist, but we're not calling `matches` to exclude it. We faced a very similar stack trace in https://github.com/opensearch-project/OpenSearch/pull/11954, where the resolution was to fix the two-phase scoring. I remember you did a fix a while back, which included adding `TwoPhase` to `HybridQueryScorer`. 

I am worried that there's still somehow a code path where it's iterating with a `TwoPhaseIterator` but thinks it's got a `DocIdSetIterator`, so it's not calling `matches`. I'm not super-familiar with the hybrid query code, but I'm wondering about the difference between `hybridScores()` and `score()`, and whether [this line](https://github.com/opensearch-project/neural-search/blame/94128848eb9e019cc8b2aa992bffc81ce71f268e/src/main/java/org/opensearch/neuralsearch/query/HybridQueryScorer.java#L190) should be `DisiWrapper topList = getSubMatches()`?

> I'll initiate a discussion on the Lucene side about a proper fix for MultiCollector.

The "incomplete" part is probably what we want to preserve. In particular, the various `ScoreMode`s are (almost) the various combinations of `isExhaustive=true/false` and `needsScores=true/false` -- the exception is `TOP_SCORES` and `TOP_DOCS_WITH_SCORES`, which are both `(false,true)`.

I wonder if it's safe enough to mostly treat the `MultiCollector` logic as taking the logical OR for these two parameters. So, e.g if any are `COMPLETE (true,true)`, you end up with `COMPLETE`. If you combine `COMPLETE_NO_SCORES(true,false)` with `TOP_SCORES(false,true)`, you get `COMPLETE(true,true)`. If you combine `TOP_DOCS(false,false)` with anything, you get the other thing, etc. 

I could be mistaken, but I think `TOP_SCORES` has some optimizations/shortcuts that we can't/don't take with `TOP_DOCS_WITH_SCORES`, so a conflict between those two should probably pick `TOP_DOCS_WITH_SCORES`.

Anyway, if the empty context uses `TOP_DOCS`, it won't "add" anything to the score mode, since that's the `(false,false)` case.

Reply:
this is a brilliant observation, especially regarding the logic in the hybrid query scorer. It appears that this was the root of the problem: two iterators (two-phase and DocIdSetIterator) are pointing to different document IDs, and if used interchangeably, this leads to an EOF issue. I need to test all known scenarios that could lead to the same problem.